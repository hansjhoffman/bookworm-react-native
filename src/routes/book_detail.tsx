import React from "react";
import { Image, Platform, StyleSheet, Text, View } from "react-native";
import { NavigationContainerProps } from "react-navigation";
import gql from "graphql-tag";
import { useQuery } from "@apollo/react-hooks";

import { Error, Loading, Stars } from "../components";
import { Colors, Layout } from "../constants";
import { Book } from "../types";
import { Images } from "../../assets";

export const BOOK_QUERY = gql`
  query bookDetail($id: ID!) {
    book(id: $id) {
      id
      authors
      description
      rating
      thumbnail
      title
    }
  }
`;

export function BookDetail({ navigation }: NavigationContainerProps) {
  const id: string = navigation && navigation.getParam("id", "");

  const { data, error, loading } = useQuery(BOOK_QUERY, { variables: { id } });

  if (loading) return <Loading />;
  if (error) return <Error message={error.message} />;

  const book: Book = data.book;

  return (
    <View style={styles.container}>
      <View style={styles.card}>
        <Image
          defaultSource={Images.placeholder}
          resizeMode="contain"
          source={{ uri: book.thumbnail || "WHY????" }}
          style={styles.thumbnail}
        />
        <View style={styles.infoContainer}>
          <Text style={styles.bookTitleText} numberOfLines={2}>
            {book.title}
          </Text>
          <Text style={styles.bookAuthorText} numberOfLines={1}>
            {book.authors}
          </Text>
          <Stars rating={book.rating} />
          <Text style={styles.bookDescriptionText} numberOfLines={4}>
            {book.description}
          </Text>
        </View>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  bookAuthorText: {
    color: Colors.textLight,
    fontFamily: "Lato-Light",
    fontSize: 14,
    marginBottom: Layout.margin_sm,
  },
  bookDescriptionText: {
    color: Colors.textMedium,
    fontFamily: "Lato-Regular",
    fontSize: 16,
    lineHeight: 25,
    marginTop: Layout.margin_sm,
    textAlign: "justify",
  },
  bookTitleText: {
    color: Colors.textDark,
    fontFamily: "Lato-Bold",
    fontSize: 20,
    marginBottom: Layout.margin_sm,
  },
  card: {
    alignItems: "center",
    backgroundColor: Colors.white,
    borderRadius: Layout.cardRadius,
    height: Layout.screen.height * 0.5,
    marginHorizontal: Layout.margin_xl,
    padding: Layout.margin_lg,
    position: "relative",
    ...Platform.select({
      android: {
        elevation: 5,
      },
      ios: {
        shadowColor: Colors.shadow,
        shadowOffset: {
          height: 7,
          width: 0,
        },
        shadowOpacity: 0.1,
        shadowRadius: Layout.cardRadius,
      },
    }),
  },
  container: {
    flex: 1,
    backgroundColor: Colors.white,
    justifyContent: "center",
  },
  icon: {
    height: 25,
    width: 25,
    marginLeft: Layout.margin_lg,
    resizeMode: "contain",
    tintColor: Colors.iconSelected,
  },
  infoContainer: {
    paddingTop: 150,
  },
  thumbnail: {
    borderRadius: 15,
    height: 300,
    overflow: "hidden",
    position: "absolute",
    top: -150,
    width: 200,
  },
});
