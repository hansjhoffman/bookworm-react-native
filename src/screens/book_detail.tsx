import React from "react";
import { Text, View } from "react-native";
import { NavigationContainerProps } from "react-navigation";

export function BookDetail({ navigation }: NavigationContainerProps) {
  const id = navigation && navigation.getParam("id", "");

  return id ? (
    <View>
      <Text>{`Details for book: ${id}`}</Text>
    </View>
  ) : (
    <View>
      <Text>Unable to find book</Text>
    </View>
  );
}
