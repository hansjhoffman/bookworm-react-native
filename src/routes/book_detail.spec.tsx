import React from "react";
import { render, wait, waitForElement } from "@testing-library/react-native";
import { MockedProvider } from "@apollo/react-testing";
import { GraphQLError } from "graphql";

import { BookDetail, BOOK_QUERY } from "./book_detail";

describe("[routes] book_detail", function() {
  let mockBook;

  it("should handle loading state", function() {
    const { getByText } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <BookDetail />
      </MockedProvider>,
    );

    getByText("Loading...");
  });

  it("should handle error", async function() {
    mockBook = {
      request: {
        query: BOOK_QUERY,
      },
      result: {
        errors: [new GraphQLError("Custom error!")],
      },
    };

    const { getByText } = render(
      <MockedProvider mocks={[mockBook]} addTypename={false}>
        <BookDetail />
      </MockedProvider>,
    );

    await waitForElement(() => getByText("GraphQL error: Custom error!"));
  });

  it("should match snapshot", async function() {
    mockBook = {
      request: {
        query: BOOK_QUERY,
        variables: {
          id: "1",
        },
      },
      result: {
        data: {
          book: {
            id: "1",
            authors: "Ayn Rand",
            description:
              "Peopled by larger-than-life heroes and villains, charged with towering \
                          questions of good and evil, Atlas Shrugged is Ayn Rand’s magnum opus: \
                          a philosophical revolution told in the form of an action \
                          thriller—nominated as one of America’s best-loved novels by PBS’s \
                          The Great American Read.",
            rating: 3,
            thumbnail: "https://images-na.ssl-image.jpg",
            title: "Atlas Shrugged",
          },
        },
      },
    };

    const withNav: any = {
      navigation: {
        getParam: jest.fn(),
        state: {
          params: { id: "1" },
        },
      },
    };

    const { baseElement } = render(
      <MockedProvider mocks={[mockBook]} addTypename={false}>
        <BookDetail {...withNav} />
      </MockedProvider>,
    );

    await wait(() => expect(baseElement).toMatchSnapshot());
  });
});
