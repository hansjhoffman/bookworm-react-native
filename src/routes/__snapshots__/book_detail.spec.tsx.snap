// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[routes] book_detail should match snapshot 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text>
      Network error: No more mocked responses for the query: query bookDetail($id: ID!) {
  book(id: $id) {
    id
    authors
    description
    rating
    thumbnail
    title
  }
}
, variables: {}
    </Text>
  </View>
</View>
`;
